File: diameterOfBinaryTree.py
# problem: https: // leetcode.com/problems/diameter-of-binary-tree/
# Runtime: 40 ms, faster than 87.56 % of Python3 online submissions for Diameter of Binary Tree.
# Memory Usage: 16 MB, less than 79.82 % of Python3 online submissions for Diameter of Binary Tree.

# Definition for a binary tree node.


class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right


class Solution:

    def traverse(self, node):
        if not node:
            return 0

        right = self.traverse(node.right)
        left = self.traverse(node.left)
        self.max = max(self.max, right + left)
        return max(right + 1, left + 1)

    def diameterOfBinaryTree(self, root: TreeNode) -> int:
        self.max = 0
        self.traverse(root)
        return self.max

File: diameterOfBinaryTree.go
// problem: https://leetcode.com/problems/diameter-of-binary-tree/
// Runtime: 4 ms, faster than 90.55% of Go online submissions for Diameter of Binary Tree.
// Memory Usage: 4.5 MB, less than 87.01% of Go online submissions for Diameter of Binary Tree.

package algorithms

//   Definition for a binary tree node.
type TreeNode struct {
	Val   int
	Left  *TreeNode
	Right *TreeNode
}

func max(a int, b int) int {
	if a > b {
		return a
	}
	return b
}

func traverse(root *TreeNode) (int, int) {
	if root == nil {
		return 0, 0
	}
	rl, rc := traverse(root.Right)
	ll, lc := traverse(root.Left)
	return 1 + max(rl, ll), max(1+rl+ll, max(rc, lc))
}

func diameterOfBinaryTree(root *TreeNode) int {
	_, answer := traverse(root)
	return answer - 1
}
